relay-server/src/actors/project_cache.rs

```diff
@@ -542,121 +194,467 @@ impl UpdateRateLimits {
+
+        // Log garbage queue size:
+        let queue_size = self.garbage_disposal.queue_size() as f64;
+        relay_statsd::metric!(gauge(RelayGauges::ProjectCacheGarbageQueueSize) = queue_size);
+
+        metric!(timer(RelayTimers::ProjectStateEvictionDuration) = eviction_start.elapsed());
     }
+
+    fn get_or_create_project(&mut self, project_key: ProjectKey) -> &mut Project {
+        metric!(histogram(RelayHistograms::ProjectStateCacheSize) = self.projects.len() as u64);
+
+        let config = self.config.clone();
+
+        self.projects
+            .entry(project_key)
+            .and_modify(|_| {
+                metric!(counter(RelayCounters::ProjectCacheHit) += 1);
+            })
+            .or_insert_with(move || {
+                metric!(counter(RelayCounters::ProjectCacheMiss) += 1);
+                Project::new(project_key, config)
+            })
+    }
+
+    fn merge_state(&mut self, message: UpdateProjectState) {
+        let UpdateProjectState {
+            project_key,
+            state,
+            no_cache,
+        } = message;
+
+        self.get_or_create_project(project_key)
+            .update_state(state, no_cache);
+    }
+
+    async fn handle_update(&mut self, message: RequestUpdate) {
+        let RequestUpdate {
+            project_key,
+            no_cache,
+        } = message;
+
+        // Bump the update time of the project in our hashmap to evade eviction.
+        self.get_or_create_project(project_key)
+            .refresh_updated_timestamp();
+
+        let source = self.source.clone();
+        let sender = self.state_tx.clone();
+
+        tokio::spawn(async move {
+            let state = source
+                .fetch(project_key, no_cache)
+                .await
+                .unwrap_or_else(|()| Arc::new(ProjectState::err()));
+
+            let message = UpdateProjectState {
+                project_key,
+                state,
+                no_cache,
+            };
+
+            sender.send(message).ok();
+        });
+    }
+
+    fn handle_get(&mut self, message: GetProjectState, sender: ProjectSender) {
+        self.get_or_create_project(message.project_key)
+            .get_state(sender, message.no_cache);
+    }
+
+    fn handle_get_cached(&mut self, message: GetCachedProjectState) -> Option<Arc<ProjectState>> {
+        self.get_or_create_project(message.project_key)
+            .get_cached_state(false)
+    }
+
+    fn handle_check_envelope(
+        &mut self,
+        message: CheckEnvelope,
+    ) -> Result<CheckedEnvelope, DiscardReason> {
+        let project = self.get_or_create_project(message.envelope.meta().public_key());
+
+        // Preload the project cache so that it arrives a little earlier in processing. However,
+        // do not pass `no_cache`. In case the project is rate limited, we do not want to force
+        // a full reload. Fetching must not block the store request.
+        project.prefetch(false);
```

Comment:
Since this always runs the prefetch, we could move this inside of `check_envelope`. However, there's also `validate_envelope` which enqueues this differently, so I propose to keep it as-is.